### ISSUE TITLE
Clients cannot handle properly closed connection in OnOpen callback

### ISSUE THREAD

**Issue by qioalice** [lesismal/nbio#193]
Hi again :)

**Use case:** 
Let's say we should validate something in WebSocket connection ASAP. Authorization token, for example. Or another entities user should provide correctly to establish WebSocket conn. When and how could we do that?

**Requirements:**
Full modern browser compatibility (Chrome, FF, Safari) + SDK or/and libraries for specific platform/languages.

**Thoughts:**
- According to https://stackoverflow.com/questions/21762596 we cannot use HTTP response. Yes, the less amount of clients could handle HTTP response during switching protocols (WebSocket handshake) and handle it properly, but it's not the generic solution. Moreover, it's not a standard.
- WebSocket standard forces us to use WebSocket close codes ASAP. To be more precision in the time interval before first message is sent/received and after connection is established. For me, that's exactly the `OnOpen` callback exists for.

This code closes connection, yes, but it doesn't seem it to be doing it right. 
```go
u.OnOpen(func(conn *websocket.Conn) {
	_ = conn.WriteMessage(websocket.CloseMessage, []byte("UNAUTHORIZED"))
	_ = conn.Close()
})
```

Let's see what clients' results we get:
- Browser: Just a console message like `app.js:1 WebSocket connection to 'ws://localhost:3000/unprotected/ws' failed: ` Neither `close` event is emitted nor `error`. 
- Gorilla client: `2022/06/06 21:30:40 OPEN malformed HTTP response "\x88\fUNAUTHORIZED"` for test code I paste below. In the code I also trying to read the first incoming WebSocket message from the server, but the connection is closed and error is returned during connection opening process. It's not even close to read the message.

Gorilla test client code:
```go
	conn, resp, err := websocket.DefaultDialer.Dial(URL, nil)
	if resp != nil {
		data, _ := ioutil.ReadAll(resp.Body)
		log.Println(resp.Status, resp.Close, string(data))
	}
	if err != nil {
		log.Fatalln("OPEN", err)
	}

	mt, data, err := conn.ReadMessage()
	if err != nil {
		log.Fatalln("READ MESSAGE", err)
	}

	log.Println("GOT MESSAGE", mt, string(data))
```

So.
How and when I may achieve my goal the right way? Am I doing something wrong or it's a bug?

PS. Again. I have to check whether user provides enough and correct data before first message is sent to/from client.

**Comment by lesismal** [comment]
I think we should move this
https://github.com/lesismal/nbio/blob/master/nbhttp/websocket/upgrader.go#L293
to
https://github.com/lesismal/nbio/blob/master/nbhttp/websocket/upgrader.go#L311

**Comment by lesismal** [comment]
Actually we can implement `OnOpen` logic after `Upgrade`:
```golang
conn, err := upgrader.Upgrade(w, r, nil)
// OnOpen logic
```

Because nbio.Engine provides `OnOpen` handler, to keep the same design, websocket provides `OnOpen` handler too.
But this should be a bug and we may fix it as I mentioned above, just need more test.

**Comment by qioalice** [comment]
I prefer your first solution that moves `OnOpen` call after last `server -> client` handshake message is sent. 
I'm afraid the second variant isn't suitable for me, because there's no guarantee that such logic would be executed before first message is received from client. 

I need a guarantee, that I can put logic in such place where it would be executed before first call of `OnMessage` handler, no matter whether client sent a message already or not.

**Comment by lesismal** [comment]
Fixed and please try the new version: https://github.com/lesismal/nbio/releases/tag/v1.2.20

The second solution to implement `OnOpen` logic after `Upgrade` can also guarantee.

**Comment by qioalice** [comment]
Sorry for the delay.
Thank you for the release. I close this issue.

TLDR about below:
Thoughts about guarantees. You don't have to comment or answer to it, but I'd be glad if you do. 

Honestly, I doesn't quite understand how exactly nbio works (I mean, how these guarantees are provided). I don't have much time to go deep to the code to figure out. If it really works like you said and we're both talking the same, you should subscribe to kqueue/epoll not right after Upgrader is done but also wait some logic placed in HTTP handler to complete. Like:

--- HTTP handler start ---
1. `Upgrader.Upgrade()`
2. <Some OnOpen logic here>
3. Start receiving messages here...
--- HTTP handler end ---
... or here?

In both of these cases, you don't have control, since in HTTP handler, AFAIK, there's no need to do something else but just call `Upgrade`. So, how could you provide a guarantee for me, that there would no data race or interceptions. And I wouldn't receive any message from client before HTTP handler is over? (I guess, you meaning this telling me I could place `OnOpen` logic after `Upgrade` in the HTTP handler).

**Comment by lesismal** [comment]
1. Every Conn has a execute queue/list.
2. All the events(onOpen, onClose) and messages(onData: http request, ws message) will be enqueued to this queue/list with locking operation to guarantee order: 
https://github.com/lesismal/nbio/blob/master/conn.go#L63
https://github.com/lesismal/nbio/blob/master/conn.go#L107
